pkg/internal/controller/controller.go

```diff
@@ -208,6 +221,11 @@ func (c *Controller) Start(ctx context.Context) error {
 	}
 
 	<-ctx.Done()
+	// decrement counts
+	for obj, count := range refCounts {
```

Comment:
Should do it on the source, probably -- sources themselves are runnables, so the source shutting down can manage the removal of the handle, and the controller doesn't need to know anything about it.  Should try to avoid breaking the abstraction if possible.

And yeah, not all controller watches are going to be Kind watches -- the intention was that they could be anything.